src/Google.Api.Gax/Platform.cs

```diff
@@ -167,37 +170,163 @@ public GaePlatformDetails(string gcloudProject, string gaeInstance, string gaeSe
     /// </summary>
     public sealed class GkePlatformDetails
     {
+        /// <summary>
+        /// Data from the Kubernetes API
+        /// </summary>
+        public class KubernetesData
+        {
+            /// <summary>
+            /// JSON from https://kubernetes/api/v1/namespaces/{namespace}
+            /// </summary>
+            public string NamespaceJson { get; set; }
+
+            /// <summary>
+            /// JSON from https://kubernetes/api/v1/namespaces/{namespace}/pods/{pod-name}
+            /// </summary>
+            public string PodJson { get; set; }
+
+            /// <summary>
+            /// Lines from /proc/self/mountinfo
+            /// </summary>
+            public string[] MountInfo { get; set; }
+        }
+
         private static readonly PathTemplate s_zoneTemplate = new PathTemplate("projects/*/zones/*");
 
+#if NETSTANDARD1_3
+        internal static async Task<KubernetesData> LoadKubernetesDataAsync()
+        {
+            var kubernetesServiceHost = Environment.GetEnvironmentVariable("KUBERNETES_SERVICE_HOST");
+            int.TryParse(Environment.GetEnvironmentVariable("KUBERNETES_SERVICE_PORT"), out var kubernetesServicePort);
+            var podName = Environment.GetEnvironmentVariable("HOSTNAME");
+            if (string.IsNullOrEmpty(kubernetesServiceHost) || kubernetesServicePort == 0 || string.IsNullOrEmpty(podName))
+            {
+                // Not running on GKE
+                return null;
+            }
+            var baseUrl = $"https://{kubernetesServiceHost}:{kubernetesServicePort}/api/v1";
+            string kubernetesNamespace;
+            string kubernetesToken;
+            string kubernetesCa;
+            string[] mountInfo;
+            try
+            {
+                kubernetesNamespace = File.ReadAllText("/var/run/secrets/kubernetes.io/serviceaccount/namespace");
+                kubernetesToken = File.ReadAllText("/var/run/secrets/kubernetes.io/serviceaccount/token");
+                kubernetesCa = File.ReadAllText("/var/run/secrets/kubernetes.io/serviceaccount/ca.crt");
+                mountInfo = File.ReadAllLines("/proc/self/mountinfo");
+            }
+            catch
+            {
+                // Various errors can occur, all of which mean probably not running on GKE
+                return null;
+            }
+            if (string.IsNullOrEmpty(kubernetesNamespace) || string.IsNullOrEmpty(kubernetesToken) ||
+                string.IsNullOrEmpty(kubernetesCa) || mountInfo == null || mountInfo.Length == 0)
+            {
+                // These files should contain useful data on GKE
+                return null;
+            }
+
+            var handler = new HttpClientHandler
+            {
+                // TODO: Properly check that the certificate is valid, using kubernetesCa
```

Comment:
This sounds like an important TODO - are we happy to ship a release without it?

Reply:
done